src/lib.rs

```diff
@@ -175,379 +201,101 @@ impl std::fmt::Debug for Step {
-            }
-            dive_mut(MutPaths {
-                starting_point: value,
-                query: remaining_query,
-            })
-        }
-        (step, value) => {
-            let step = step.name();
-            let value = value_name(value);
-            tracing::warn!("'{step}' not supported for '{value}'",);
-
-            DiveOutcome::Nothing
+    let mut current_node = Some(node);
+    for fragment in &adr.0 {
+        let actual_node = current_node?;
+        match fragment {
+            Field(f) if f == "." => current_node = Some(actual_node),
+            Field(f) => current_node = actual_node.get(f),
+            Index(i) => current_node = actual_node.get(i),
         }
     }
-}
 
-fn dive(path: Paths<'_>) -> DiveOutcome<'_, Paths> {
-    let Some((next_step, remaining_query)) = path.query.take_step() else {
-        return DiveOutcome::Final(path.starting_point);
-    };
-
-    tracing::info!("Matching {next_step:?} against {:?}", path.starting_point);
-    match (next_step, path.starting_point) {
-        (Step::Field(f), Value::Mapping(m)) => {
-            let accessor = Value::String(f);
-            let Some(next_value) = m.get(accessor) else {
-                return DiveOutcome::Nothing;
-            };
-            dive(Paths {
-                starting_point: next_value,
-                query: remaining_query,
-            })
-        }
-        (Step::At(idx), Value::Sequence(s)) => {
-            let Some(next_value) = s.get(idx) else {
-                return DiveOutcome::Nothing;
-            };
-            dive(Paths {
-                starting_point: next_value,
-                query: remaining_query,
-            })
-        }
-        (Step::Range(r), Value::Sequence(sequence)) => {
-            let mut additional_paths = Vec::new();
-            for point in &sequence[r] {
-                additional_paths.push(Paths {
-                    starting_point: point,
-                    query: remaining_query.clone(),
-                });
-            }
-            DiveOutcome::Branch(additional_paths)
-        }
-        (Step::All, Value::Sequence(sequence)) => {
-            let mut additional_paths = Vec::new();
-            for point in sequence {
-                additional_paths.push(Paths {
-                    starting_point: point,
-                    query: remaining_query.clone(),
-                });
-            }
-            DiveOutcome::Branch(additional_paths)
-        }
-        (Step::Filter(field, predicate), s @ Value::String(_)) => {
-            if field != "." {
-                return DiveOutcome::Nothing;
-            }
+    current_node
+}
 
-            if !predicate(s) {
-                return DiveOutcome::Nothing;
-            }
+pub struct ManyMutResults<'input> {
+    candidates: VecDeque<Candidate>,
+    // the addresses here have to be relative to the root
+    found_addresses: VecDeque<Address>,
+    root_node: &'input mut Value,
```

Comment:
Put a `#[pint_project]` on it to not move it.
this might solve the lifetime rubber-banding with unsafe below